File: src/api/core/scheduler.ts
import symbols from "../ffi/ffi.ts";

/**
 * Apply linear decay scheduling for learning rate.
 * @param rate Rate of decay
 * @param step_size Number of steps between each decay.
 */
export function linearDecay(
  rate: number,
  step_size: number,
): Deno.PointerValue {
  return symbols.linear_decay_scheduler(rate, step_size);
}
/**
 * Apply exponential decay scheduling for learning rate.
 * @param rate Rate of decay
 * @param step_size Number of steps between each decay.
 */
export function exponentialDecay(
  rate: number,
  step_size: number,
): Deno.PointerValue {
  return symbols.exponential_decay_scheduler(rate, step_size);
}
/**
 * Apply one cycle scheduling for learning rate.
 * Make sure to set learning rate to a very small value while training.
 * @param max_rate Maximum value the learning rate can get to.
 * @param step_size Number of steps between each decay.
 */
export function oneCycle(
  max_rate: number,
  step_size: number,
): Deno.PointerValue {
  return symbols.one_cycle_scheduler(max_rate, step_size);
}

/**
 * Apply no scheduling
 */
export function noDecay(): Deno.PointerValue {
  return symbols.no_decay_scheduler();
}

File: crates/classy/src/core/scheduler.rs
pub enum Scheduler {
    None,
    LinearDecay {
        // Rate of decay
        rate: f32,
        // Number of epochs for decay
        step_size: usize,
    },
    ExponentialDecay {
        // Rate of decay
        rate: f32,
        // Number of epochs for decay
        step_size: usize,
    },
    OneCycle {
        // Max allowed learning rate
        max_rate: f32,
        // Number of steps in one cycle
        step_size: usize,
    },
}

impl Scheduler {
    pub fn eta(&self, learning_rate: f32, step: usize) -> f32 {
        match self {
            Scheduler::None => learning_rate,
            Scheduler::LinearDecay { rate, step_size } => {
                learning_rate / (rate * (1 + step / step_size) as f32)
            }
            Scheduler::ExponentialDecay { rate, step_size } => {
                learning_rate * rate.powi((step / step_size) as i32)
            }
            Scheduler::OneCycle {
                max_rate,
                step_size,
            } => {
                let steps = *step_size as f32;
                let step = step % (2 * step_size);
                if step < *step_size {
                    learning_rate + (max_rate - learning_rate) * (step as f32) / (steps)
                } else {
                    max_rate - (max_rate - learning_rate) * ((step - step_size) as f32) / (steps)
                }
            }
        }
    }
}
